DOCSP-41995: transaction (#137)
mongodb/docs-php-library: source/write/transaction.txt

.. _php-transactions:

=====================
Perform a Transaction
=====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, ACID compliance, multi-document

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use the {+driver-short+} to perform
**transactions**. Transactions allow you to perform a series of operations
that change data only if the entire transaction is committed.

> If any operation in the transaction does not succeed, the driver stops the

Reviewer: S: I think "library" is a better word than "driver" for PHP
```suggestion
If any operation in the transaction does not succeed, the library stops the
```

by using the methods provided by the ``Session`` class. The
following table describes these methods:

.. list-table::
   :widths: 25 75
   :stub-columns: 1
   :header-rows: 1

   * - Method
     - Description

   * - ``startTransaction()``
     - | Starts a new transaction on this session. The session
         must be passed into each operation within the transaction, or
         the operation will run outside of the transaction.
       |
       | You can set transaction options by passing an options parameter.

   * - ``commitTransaction()``
     - | Commits the active transaction for this session. This method returns an
         error if there is no active transaction for the session or the
         transaction was previously ended.

   * - ``abortTransaction()``
     - | Ends the active transaction for this session. This method returns an
         error if there is no active transaction for the session or if the
         transaction was committed or ended.

.. _php-txn-example:

Transaction Example
-------------------

The following code defines the ``transferTxn()`` callback function that
modifies data in the ``saving_accounts``, ``checking_accounts``, and ``receipts``

> collections of the ``bank`` database:

Reviewer: S: not a necessary change, but there's a lot going on in this code example and it might be helpful to give more detail about how you're modifying those collections in a bulleted list

Author: good point!

.. literalinclude:: /includes/write/transaction.php
   :language: php
   :dedent:
   :start-after: begin-callback
   :end-before: end-callback

Then, run the following code to perform the transaction. This code
completes the following actions:

1. Creates a session from the client by using the ``startSession()`` method.
#. Calls the ``with_transaction()`` method to manage the transaction,
   passing the session and the callback as parameters.

> .. io-code-block::
> 

Reviewer: S: make this code example copyable
```suggestion
.. io-code-block::
   :copyable:
```

   .. input:: /includes/write/transaction.php
      :language: php
      :dedent:
      :start-after: begin-txn
      :end-before: end-txn

   .. output:: 
      :language: console
      :visible: false

      Successfully performed transaction!

Additional Information
----------------------

To learn more about the concepts mentioned in this guide, see the
following pages in the {+mdb-server+} manual:

- :manual:`Transactions </core/transactions/>`
- :manual:`Server Sessions </reference/server-sessions/>`

> - :manual:`Read Isolation, Consistency, and Recency
>   </core/read-isolation-consistency-recency/#causal-consistency>`

Reviewer: Q: did you intentionally link to the Casual Consistency section of this page? I think you can just link to the main page
```suggestion
- :manual:`Read Isolation, Consistency, and Recency
  </core/read-isolation-consistency-recency/>`
```

To learn more about ACID compliance, see the :website:`What are ACID
Properties in Database Management Systems? </basics/acid-transactions>`
article on the MongoDB website.

To learn more about insert operations, see the
:ref:`php-write-insert` guide.

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about the methods and types mentioned in this
guide, see the following API documentation:

- :phpclass:`MongoDB\Client`
- :phpmethod:`MongoDB\Client::startSession()`
- :phpmethod:`MongoDB\Collection::updateOne()`
- :phpmethod:`MongoDB\Collection::insertOne()`

To learn more about the ``Session`` class and methods,
see the following {+extension-short+} API documentation:

- `MongoDB\\Driver\\Session

>   <https://www.php.net/manual/en/class.mongodb-driver-session.php>`__

Reviewer: S: you can use the {+php-manual+} constant in these links
```suggestion
  <{+php-manual+}/class.mongodb-driver-session.php>`__
```

Reviewer: applies to all

**transactions**. Transactions allow you to perform a series of operations
that change data only if the entire transaction is committed.
If any operation in the transaction does not succeed, the library stops the
transaction and discards all data changes before they ever become
visible. This feature is called **atomicity**.

In MongoDB, transactions run within logical sessions. A
session is a grouping of related read or write operations that you
want to run sequentially. Sessions enable causal consistency for a group
of operations and allow you to run operations in an **ACID-compliant**
transaction, which is a transaction that meets an expectation of
atomicity, consistency, isolation, and durability. MongoDB guarantees
that the data involved in your transaction operations remains
consistent, even if the operations encounter unexpected errors.

When using the {+php-library+}, you can create a new session from a
``MongoDB\Client`` instance. Then, you can use the resulting
``MongoDB\Driver\Session`` instance to perform transactions. You can
improve your app's performance by reusing your client for multiple
sessions and transactions instead of instantiating a new client each

> time.

Reviewer 2: It's very uncommon that a PHP app would would create multiple clients. Also, it's quite possible that multiple Clients could share the same underlying libmongoc client anyway if they're constructed with the same arguments (see: [Connection handling and persistence](https://www.php.net/manual/en/mongodb.connection-handling.php)).

I don't think we need this sentence.

.. warning::

   Use a ``Session`` only in operations running on the
   ``Client`` that created it. Using a ``Session`` with a
   different ``Client`` results in operation errors.

Methods
-------

Create a ``Session`` by using the ``MongoDB\Client::startSession()``
method on your ``Client`` instance. The {+php-library+} provides a

> Convenient Transaction API to manage the transaction lifecyle. Use the

Reviewer 2: I noticed the name is emphasized in the Node docs: https://www.mongodb.com/docs/drivers/node/current/usage-examples/transaction-conv/

Should that be done here?

> ``MongoDB\with_transaction()`` method to run custom callback within a

Reviewer 2: ```suggestion
``MongoDB\with_transaction()`` function to run custom callback within a
```

Technical nit: method is a function on a class. This is a standalone utility function not tied to any class.

Author: replaced in all instances

transaction. The ``with_transaction()`` method starts the transaction,

> then either commits it or ends it if there are errors. The

Reviewer 2: ```suggestion
then either commits it or ends it if there are errors. The
```

This sentence overlooks the actual benefit of the Convenient Transaction API. See: https://github.com/mongodb/specifications/blob/master/source/transactions-convenient-api/transactions-convenient-api.md#abstract

We should mention that it's capable of automatically retrying the transaction for certain errors such as a transient write conflict.

Author: revised this section

:ref:`php-txn-example` section of this guide demonstrates how to use
this API to perform a transaction. 

Alternatively, you can have more control over your transaction lifecyle
by using the methods provided by the ``Session`` class. The
following table describes these methods:

.. list-table::
   :widths: 25 75
   :stub-columns: 1
   :header-rows: 1

   * - Method
     - Description

   * - ``startTransaction()``
     - | Starts a new transaction on this session. The session
         must be passed into each operation within the transaction, or
         the operation will run outside of the transaction.
       |
       | You can set transaction options by passing an options parameter.

   * - ``commitTransaction()``
     - | Commits the active transaction for this session. This method returns an
         error if there is no active transaction for the session or the

>          transaction was previously ended.

Reviewer 2: Notably, a commit can fail due to other transient errors, such as a write conflict. That would be more common as a runtime error, vs. a user calling this when there is no active transaction (likely a poorly written application, and a mistake they'd make once and correct).

Author: added that clause

   * - ``abortTransaction()``
     - | Ends the active transaction for this session. This method returns an
         error if there is no active transaction for the session or if the
         transaction was committed or ended.

.. _php-txn-example:

Transaction Example
-------------------

This example defines a callback function that
modifies data in the collections of the ``bank`` database for a
banking transaction. The code performs the following actions:

- Creates ``Collection`` instances to access the target
  collections.
- Specifies the account number and amount to be transferred between
  accounts.

> - Defines the callback function, passing the ``Session`` instance as a

Reviewer 2: ```suggestion
- Defines the callback function, which receives a ``Session`` instance as a
```

Technically, the `with_transaction()` function passes the parameter each time it invokes the callback. This is the same Session that the application passes into `with_transaction()`.

.. io-code-block::
   :copyable:

   .. input:: /includes/write/transaction.php
      :language: php
      :dedent:
      :start-after: begin-txn
      :end-before: end-txn
      :emphasize-lines: 1, 4

   .. output:: 
      :language: console
      :visible: false

      Successfully performed transaction!
      Summary: SAVINGS +1000 CHECKING -1000

Additional Information
----------------------

To learn more about the concepts mentioned in this guide, see the
following pages in the {+mdb-server+} manual:

- :manual:`Transactions </core/transactions/>`
- :manual:`Server Sessions </reference/server-sessions/>`
- :manual:`Read Isolation, Consistency, and Recency
  </core/read-isolation-consistency-recency/>`

To learn more about ACID compliance, see the :website:`What are ACID
Properties in Database Management Systems? </basics/acid-transactions>`
article on the MongoDB website.

To learn more about insert operations, see the
:ref:`php-write-insert` guide.

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about the methods and types mentioned in this
guide, see the following API documentation:

- :phpclass:`MongoDB\Client`
- :phpmethod:`MongoDB\Client::startSession()`

> - :phpmethod:`MongoDB\with_transaction()`

Reviewer 2: I'll note that the docs use a common `:phpmethod:` directive here, even though this is technically a function. Nothing to address here -- just wanted to point that out.

transaction. The ``with_transaction()`` function performs the following
tasks:

- Starts the transaction
- Handles errors by either ending the transaction or retrying it, such
  as when the operation returns a ``TransientTransactionError``
- Commits the transaction

The :ref:`php-txn-example` section of this guide demonstrates how to use
this API to perform a transaction. 

.. _php-core-txn:

Core API
~~~~~~~~

Alternatively, you can have more control over your transaction lifecyle
by using the methods provided by the ``Session`` class. The
following table describes these methods:

.. list-table::
   :widths: 25 75
   :stub-columns: 1
   :header-rows: 1

   * - Method
     - Description

   * - ``startTransaction()``
     - | Starts a new transaction on this session. The session
         must be passed into each operation within the transaction, or
         the operation will run outside of the transaction.
       |
       | You can set transaction options by passing an options parameter.

   * - ``commitTransaction()``
     - | Commits the active transaction for this session. This method returns an
         error if there is no active transaction for the session, the

>          transaction was previously ended, or if there is a write
>          conflict.

Reviewer 2: ```suggestion
         transaction was previously ended, or if there is a write conflict.
```

If you want to avoid extra wrapping.
